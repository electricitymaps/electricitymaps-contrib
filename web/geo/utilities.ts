import * as fs from 'node:fs';

import {
  area,
  Feature,
  FeatureCollection,
  featureCollection,
  featureEach,
  getCoords,
  getType,
  length,
  MultiPolygon,
  Polygon,
  polygon,
  polygonToLineString,
  Properties,
  truncate,
} from '@turf/turf';

/* Transform the feature collection of polygons and multi-polygons into a feature collection of polygons only */
/* all helper functions should rely on its output */
const handlePolygon = (feature: Feature<Polygon, Properties>) =>
  polygon(getCoords(feature), feature.properties);

const handleMultiPolygon = (feature: Feature<MultiPolygon, Properties>) =>
  getCoords(feature).map((coord) => polygon(coord, feature.properties));

function getPolygons(fc: FeatureCollection<Polygon | MultiPolygon, Properties>) {
  const polygons: Feature<Polygon, Properties>[] = [];

  featureEach(fc, (feature) => {
    const type = getType(feature);
    switch (type) {
      case 'Polygon': {
        polygons.push(handlePolygon(feature as Feature<Polygon, Properties>));
        break;
      }
      case 'MultiPolygon': {
        polygons.push(
          ...handleMultiPolygon(feature as Feature<MultiPolygon, Properties>)
        );
        break;
      }
      default: {
        throw new Error(`Encountered unhandled type: ${type}`);
      }
    }
  });

  return truncate(featureCollection(polygons), { precision: 6 });
}

function isSliver(
  polygon: Feature<Polygon, Properties>,
  polArea: number,
  sliverRatio: number
) {
  const lineStringLength = length(polygonToLineString(polygon));
  return Number(lineStringLength / polArea) > Number(sliverRatio);
}

function getHoles(
  fc: FeatureCollection<Polygon | MultiPolygon, Properties>,
  minArea: number,
  sliverRatio: number
) {
  const holes: Feature<Polygon, Properties>[] = [];
  featureEach(fc, (feature: Feature<Polygon | MultiPolygon, Properties>) => {
    const coords = getCoords(feature);
    if (coords.length > 0) {
      for (const coord of coords) {
        const pol = polygon([coord]);
        const polArea = area(pol);
        if (polArea < minArea && !isSliver(pol, polArea, sliverRatio)) {
          holes.push(pol);
        }
      }
    }
  });
  return featureCollection(holes);
}

const fileExists = (fileName: string) => fs.existsSync(fileName);

const getJSON = (fileName: string) => JSON.parse(fs.readFileSync(fileName, 'utf8'));

const writeJSON = (fileName: string, object: unknown) =>
  fs.writeFileSync(fileName, JSON.stringify(object), { encoding: 'utf8', flag: 'w' });

function log(message: string) {
  console.error('\u001B[31m%s\u001B[0m', `ERROR: ${message}`);
}

/**
 * Function to round a number to a specific amount of decimals.
 * @param {number} number - The number to round.
 * @param {number} decimals - Defaults to 2 decimals.
 * @returns {number} Rounded number.
 */
const round = (number: number, decimals = 2): number =>
  Math.round((number + Number.EPSILON) * 10 ** decimals) / 10 ** decimals;

export {
  fileExists,
  getHoles,
  getJSON,
  getPolygons,
  handleMultiPolygon,
  handlePolygon,
  isSliver,
  log,
  round,
  writeJSON,
};
