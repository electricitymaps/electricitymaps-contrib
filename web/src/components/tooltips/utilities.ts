import { SIDEBAR_WIDTH } from 'features/app-sidebar/AppSidebar';

export const getSafeTooltipPosition = (
  mousePositionX: number,
  mousePositionY: number,
  screenWidth: number,
  tooltipWidth: number,
  tooltipHeight: number
) => {
  const sidebarWidth = Number.parseInt(SIDEBAR_WIDTH.replace('px', ''));
  const ToolTipFlipBoundaryX = tooltipWidth + 80;
  const ToolTipFlipBoundaryY = tooltipHeight + 16;
  const xOffset = sidebarWidth + 10;
  const yOffset = tooltipHeight + 16;

  const tooltipPosition = {
    x: mousePositionX + xOffset,
    y: mousePositionY - yOffset,
  };
  if (screenWidth - mousePositionX < ToolTipFlipBoundaryX) {
    tooltipPosition.x = mousePositionX - tooltipWidth + sidebarWidth - 10;
  }
  if (mousePositionY < ToolTipFlipBoundaryY) {
    tooltipPosition.y = mousePositionY;
  }

  return tooltipPosition;
};

export const getOffsetTooltipPosition = ({
  mousePositionX,
  mousePositionY,
  tooltipHeight,
  isBiggerThanMobile,
}: {
  mousePositionX: number;
  mousePositionY: number;
  tooltipHeight: number;
  isBiggerThanMobile: boolean;
}) => {
  const xOffset = 10;
  const yOffset = tooltipHeight - 40;

  // For smaller screens we translate the tooltip to the top
  if (!isBiggerThanMobile) {
    return {
      x: 0,
      y: 0,
    };
  }

  const tooltipPosition = {
    x: mousePositionX + xOffset,
    y: Math.max(mousePositionY - yOffset, 0),
  };

  return tooltipPosition;
};
