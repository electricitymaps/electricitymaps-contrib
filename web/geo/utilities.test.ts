import { Properties } from '@turf/turf';
import { Feature, MultiPolygon, Polygon } from 'geojson';
import { describe, expect, it, vi } from 'vitest';

import { handleMultiPolygon, handlePolygon, log } from './utilities';

describe('log', () => {
  it('should log the correct error message with formatting', () => {
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {});

    const message = 'Test error message';
    log(message);

    expect(consoleErrorSpy).toHaveBeenCalledWith(
      '\u001B[31m%s\u001B[0m',
      `ERROR: ${message}`
    );

    consoleErrorSpy.mockRestore();
  });
});

describe('handlePolygon', () => {
  it('should correctly handle a valid polygon feature', () => {
    const feature: Feature<Polygon, Properties> = {
      type: 'Feature',
      geometry: {
        type: 'Polygon',
        coordinates: [
          [
            [0, 0],
            [0, 1],
            [1, 1],
            [1, 0],
            [0, 0],
          ],
        ],
      },
      properties: {
        name: 'Test Polygon',
      },
    };

    const actual = handlePolygon(feature);

    expect(actual).toMatchSnapshot();
  });
});

describe('handleMultiPolygon', () => {
  it('should correctly handle a valid MultiPolygon feature', () => {
    const feature: Feature<MultiPolygon, Properties> = {
      type: 'Feature',
      geometry: {
        type: 'MultiPolygon',
        coordinates: [
          [
            [
              [0, 0],
              [0, 1],
              [1, 1],
              [1, 0],
              [0, 0],
            ],
          ],
          [
            [
              [2, 2],
              [2, 3],
              [3, 3],
              [3, 2],
              [2, 2],
            ],
          ],
        ],
      },
      properties: {
        name: 'Test MultiPolygon',
      },
    };

    const actual = handleMultiPolygon(feature);

    expect(actual).toMatchSnapshot();
  });
});
