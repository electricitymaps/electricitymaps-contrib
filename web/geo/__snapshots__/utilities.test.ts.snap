// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleMultiPolygon > should correctly handle a valid MultiPolygon feature 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        [
          [
            0,
            0,
          ],
          [
            0,
            1,
          ],
          [
            1,
            1,
          ],
          [
            1,
            0,
          ],
          [
            0,
            0,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": {
      "name": "Test MultiPolygon",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        [
          [
            2,
            2,
          ],
          [
            2,
            3,
          ],
          [
            3,
            3,
          ],
          [
            3,
            2,
          ],
          [
            2,
            2,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": {
      "name": "Test MultiPolygon",
    },
    "type": "Feature",
  },
]
`;

exports[`handlePolygon > should correctly handle a valid polygon feature 1`] = `
{
  "geometry": {
    "coordinates": [
      [
        [
          0,
          0,
        ],
        [
          0,
          1,
        ],
        [
          1,
          1,
        ],
        [
          1,
          0,
        ],
        [
          0,
          0,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": {
    "name": "Test Polygon",
  },
  "type": "Feature",
}
`;
